src/math/mat4.h

```diff
@@ -0,0 +1,445 @@
+                     m_MatrixElements[12] * m_MatrixElements[5]  * m_MatrixElements[11] -
+                     m_MatrixElements[12] * m_MatrixElements[7]  * m_MatrixElements[9];
+
+        minors[3] = -m_MatrixElements[4]  * m_MatrixElements[9]  * m_MatrixElements[14] +
+                     m_MatrixElements[4]  * m_MatrixElements[10] * m_MatrixElements[13] +
+                     m_MatrixElements[8]  * m_MatrixElements[5]  * m_MatrixElements[14] -
+                     m_MatrixElements[8]  * m_MatrixElements[6]  * m_MatrixElements[13] -
+                     m_MatrixElements[12] * m_MatrixElements[5]  * m_MatrixElements[10] +
+                     m_MatrixElements[12] * m_MatrixElements[6]  * m_MatrixElements[9];
+
+        return m_MatrixElements[0] * minors[0] + m_MatrixElements[1] * minors[1] + m_MatrixElements[2] * minors[2] + m_MatrixElements[3] * minors[3];
+    }
+
+    constexpr Mat4 Mat4::Identity() noexcept {
+        return Mat4(1.0f);
+    }
+
+    constexpr Mat4 Mat4::Translation(const Vec3& translation) noexcept {
+        Mat4 result(1.0f);
+
+        result.m_MatrixElements[0 + 3 * 4] = translation.GetX();
+        result.m_MatrixElements[1 + 3 * 4] = translation.GetY();
+        result.m_MatrixElements[2 + 3 * 4] = translation.GetZ();
+
+        return result;
+    }
+
+    constexpr Mat4 Mat4::Scale(const Vec3& scale) noexcept {
+        Mat4 result(1.0f);
+
+        result.m_MatrixElements[0 + 0 * 4] = scale.GetX();
+        result.m_MatrixElements[1 + 1 * 4] = scale.GetY();
+        result.m_MatrixElements[2 + 2 * 4] = scale.GetZ();
+
+        return result;
+    }
+
+    constexpr Mat4 Mat4::Rotation(float angle, const Vec3& axis) noexcept {
+        Mat4 result(1.0f);
+
+        float angleInRadians = angle / 180.0f * static_cast<float>(M_PI);
+        float c = static_cast<float>(cos(angleInRadians));
+        float s = static_cast<float>(sin(angleInRadians));
+        float omc = 1 - c;
+
+        result.m_MatrixElements[0 + 0 * 4] = axis.GetX() * axis.GetX() * omc + c;
+        result.m_MatrixElements[1 + 0 * 4] = axis.GetX() * axis.GetY() * omc + axis.GetZ() * s;
+        result.m_MatrixElements[2 + 0 * 4] = axis.GetX() * axis.GetZ() * omc - axis.GetY() * s;
+
+        result.m_MatrixElements[0 + 1 * 4] = axis.GetX() * axis.GetY() * omc - axis.GetZ() * s;
+        result.m_MatrixElements[1 + 1 * 4] = axis.GetY() * axis.GetY() * omc + c;
+        result.m_MatrixElements[2 + 1 * 4] = axis.GetY() * axis.GetZ() * omc + axis.GetX() * s;
+
+        result.m_MatrixElements[0 + 2 * 4] = axis.GetX() * axis.GetZ() * omc + axis.GetY() * s;
+        result.m_MatrixElements[1 + 2 * 4] = axis.GetY() * axis.GetZ() * omc - axis.GetX() * s;
+        result.m_MatrixElements[2 + 2 * 4] = axis.GetZ() * axis.GetZ() * omc + c;
+
+        return result;
+    }
+
+    constexpr Mat4
+    Mat4::Orthographic(float left, float right, float bottom, float top, float near, float far) noexcept {
```

Comment:
there is no test for that function

Reply:
Hard to test numerically

Reply:
maybe one really simple one to check if there is no mistakes in index

Reply:
Done